src/pybind/mgr/dashboard_v2/README.rst

```diff
@@ -310,36 +416,62 @@ can be used:
+etc.), you may want to limit the time that the frontend will wait for your
+controller's response, by returning the information that no data could be
+retrieved within a time interval, or by returning stale data retrieved in
+previous requests.
 
-  import rbd
-  from ..tools import ApiController, RESTController
+This way the frontend is in control of the requests timing and can warn or
+ask for feedback from the user, when such request takes longer than expected.
 
-  @ApiController('rbdimages')
-  class RbdImages(RESTController):
-      def __init__(self):
-          self.ioctx = self.mgr.rados.open_ioctx('rbd')
-          self.rbd = rbd.RBD()
+To help the developer in handling these kinds of situations we provide a
+function decorator called ``@ViewCache``, which the developer can use to
+decorate a function that makes accesses to the Ceph cluster and returns data.
 
-      def list(self):
-          return [{'name': n} for n in self.rbd.list(self.ioctx)]
+Example::
 
-In the example above, we want to mock the return value of the ``rbd.list``
-function, so that we can test the JSON response of the controller.
+  import rbd
+  from ..tools import ApiController, RESTController, ViewCache
 
-The unit test code will look like the following::
 
-  import mock
-  from .helper import ControllerTestCase
+  @ApiController('rbd')
+  class Rbd(RESTController):
+      def __init__(self):
+          super(Rbd, self).__init__()
+          self.rbd = rbd.RBD()
 
-  class RbdImagesTest(ControllerTestCase):
-      @mock.patch('rbd.RBD.list')
-      def test_list(self, rbd_list_mock):
-          rbd_list_mock.return_value = ['img1', 'img2']
-          self._get('/api/rbdimages')
-          self.assertJsonBody([{'name': 'img1'}, {'name': 'img2'}])
+      @ViewCache()
+      def image_list(self, pool_name):
+          ioctx = self.mgr.rados.open_ioctx(pool_name)
+          return self.rbd.list(ioctx)
+
+      def list(self, pool_name):
+          state, imglist = self.image_list(pool_name)
+          if state == ViewCache.VALUE_EXCEPTION:
+              result = str(imglist)
+          elif state == ViewCache.VALUE_NONE:
+              result = None
+          elif not imglist:
+              result = []
+          else:
+              result = [{'name': n} for n in imglist] if imglist else None
+          return {'state': state, 'result': result}
+
+When decorating a function with ``@ViewCache``, that function is transformed
+to also return a state flag besides the data that was already being returned by
+the function.
+The ``@ViewCache`` accepts a timeout argument in seconds that specifies the
+time to wait for the function to return. The default timeout is 5 seconds.
+
+The state flag can have four different values:
+
+* ``VALUE_OK = 0``: the returned data is fresh.
+* ``VALUE_STALE = 1``: the returned data might be stale.
+* ``VALUE_NODE = 2``: no data could be returned.
+* ``VALUE_EXCEPTION = 3``: an exception was raised while running the function.
```

Comment:
Which integer value they exactly correspond to is somehow irrelevant.

Reply:
It's relevant for the frontend guys that consume this